### DISCUSSION TITLE
Forcing custom created/updated values

### DISCUSSION THREAD

**Discussion by DaniDipp** [pocketbase/pocketbase#1186] [Q&A] [answered]
I'm migrating data from MongoDB to Pocketbase and would like to keep my timestamps. It's not as easy as sending `created` and `updated` values with the create request. How could I accomplish this?

**Comment by adamshand** [comment]
Following.

**Comment by ganigeorgiev** [comment, marked as answer]
The `created` and `updated` fields cannot be set with the HTTP API. They can be set only if you use PocketBase as framework or when directly working with the db.

There are several ways you can apply the import but I think the easiest process probably would be:

1. Create 2 temporary `mongo_created` and `mongo_updated` `date` fields in your collection.

2. Run the import by assigning the mongo `created`/`updated` timestamp fields to the fields from 1).

3. Transfer all `mongo_created` to `created` and `mongo_updated` to `updated` by executing the following SQL:
   ```sh
   sqlite3 pb_data/data.db 'UPDATE your_collection_name SET created = mongo_created, updated = mongo_updated'
   ```

4. Verify that the `created` and `updated` are updated and remove the temporary `mongo_created` and `mongo_updated` fields from your collection in the Admin UI.

**Reply by DaniDipp** [reply]
That worked like a charm, thanks!

**Comment by silberjan** [comment]
I found a similar workaround that allows setting explicit created dates through the rest api using the `OnModelBeforeCreate` hook. 

1. Create a (temporary) `explicitCreated` date Field on your collection
2. Set the `created` field to the value of `explicitCreated` in the hook (excuse the probably bad go syntax, im very new to it):
``` go
	app.OnModelBeforeCreate().Add(func(e *core.ModelEvent) error {
		record, ok := e.Model.(*models.Record)
		explicitCreated := record.Get("explicitCreated")

		if ok && explicitCreated != nil {
			record.Set("created", explicitCreated)
		}

		return nil
	})
```
3. (Optional) Remove the explicitCreated field once all records are migrated

This does not check the name of the collection so it should work with any collection that you add the `explicitCreated` field to.

I have one more question for @ganigeorgiev though:

Is there a way to achieve this without the addition of the temporary field? I tried a lot and it seems that during serialization all json field that might have been sent in the request are discarded if they don't match the schema.

**Reply by ganigeorgiev** [reply]
In your case I think you can use the `*Request` hooks since you'll have access to the request body data.

For example (_I haven't tested it_):

```go
app.OnRecordBeforeCreateRequest().Add(func(e *core.RecordCreateEvent) error {
	requestData := apis.RequestData(e.HttpContext)

	e.Record.Set("created", requestData.Data["created"])
	e.Record.Set("updated", requestData.Data["updated"])

	return nil
})
```

**Reply by silberjan** [reply]
<img width="322" alt="image" src="https://user-images.githubusercontent.com/12273306/210141555-8d8394e3-1cb2-4dc6-948b-9c7b9d73e9bd.png">

tried it and it works! thanks a lot!

**Comment by stamler** [comment]
Is it not possible to set 'created' and 'updated' on a new record?

```go
				record := core.NewRecord(collection)
				record.Set("created", "2024-10-10 19:00:00.000Z")
				record.Set("updated", "2024-10-10 19:00:00.000Z")
				...
				txApp.Save(record)
```
When I run this code, I get `""` for both created and updated but all the other fields I set work.

**Reply by ganigeorgiev** [reply]
@stamler I'll consider documenting it better but with PocketBase v0.23+ `created` and `updated` are optional `autodate` type collection fields and their values are automanaged, aka. they are not supposed to be modified manually - if that's what you want change the field type to `date`.

In the rare cases (e.g. migrations) when you want to set a predefined `autodate` value manually you'll have to use something like:

```go
date, err := types.ParseDateTime("2024-10-10 19:00:00.000Z")
record.SetRaw("created", date)
record.SetRaw("updated", date)
```

With that said, please don't hijack old discussions and try to provide a little more context with your questions to understand the bigger picture and what actually you are trying to do.

**Reply by stamler** [reply]
Sorry, I though these were related issues which was why I responded here rather than starting a new discussion. I'll create a new discussion in the future with more specific context. I've implemented an undo delete action from serialized JSON. Your reply about the `autodate` type answered my query, however `record.SetRaw("created", date)` still leaves a blank value unfortunately. I'll keep digging. Hopefully your comments about v0.23+ help someone in the future who searches for this issue as I did so they don't have to open a new discussion.
